dcrtimed/backend/filesystem/dump.go

```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
+		// 3. Make sure all digests exist in flush record
+		if len(digests) != len(flushRecord.Hashes) {
+			// XXX Dump hashes
+			return fmt.Errorf("   *** ERROR mismatched len: %v %v",
+				len(digests), len(flushRecord.Hashes))
+		}
+		for _, v := range flushRecord.Hashes {
+			hash := hex.EncodeToString(v[:])
+			if options.PrintHashes {
+				fmt.Printf("     Hashes       : %v\n", hash)
+			}
+
+			if _, ok := digests[hash]; !ok {
+				return fmt.Errorf("   *** ERROR unknown hash: %v",
+					hash)
+			}
+		}
+
+		// 3.1 Recreate merkle and verify it
+		root := merkle.Root(flushRecord.Hashes)
+		if !bytes.Equal(root[:], flushRecord.Root[:]) {
+			return fmt.Errorf("   *** ERROR mismatched merkle "+
+				"root: %x %x", *root, flushRecord.Root)
+		}
+
+		// 3.2 verify that all digests exist in global db and verify
+		// timestamp points to the correct container.
+		for _, v := range flushRecord.Hashes {
+			gdbts, err := fs.db.Get(v[:], nil)
+			if err != nil {
+				return fmt.Errorf("   *** ERROR not found in "+
+					"db: %x", v)
+			}
+			dbts := int64(binary.LittleEndian.Uint64(gdbts))
+			if dbts != ts {
+				return fmt.Errorf("   *** ERROR timestamp "+
+					"mismatch: %v %v", dbts, ts)
+			}
+		}
+
+		// 3.3 Verify merkle root in tx
+		u := "https://alpha.dcrdata.org/api/tx/" +
+			flushRecord.Tx.String() + "/out"
+		r, err := http.Get(u)
+		if err != nil {
+			return fmt.Errorf("   *** ERROR invalid dcrdata "+
+				"answer: %v", err)
+		}
+		var txOuts []types.TxOut
+		d := json.NewDecoder(r.Body)
+		if err := d.Decode(&txOuts); err != nil {
+			return err
+		}
+
+		var done bool
+		for _, v := range txOuts {
+			if !types.IsNullDataScript(v.ScriptPubKeyDecoded.Type) {
+				continue
+			}
+			script, err := hex.DecodeString(v.ScriptPubKeyDecoded.Hex)
+			if err != nil {
+				// XXX bad error, fix
+				return fmt.Errorf("   *** ERROR invalid "+
+					"dcrdata script: %v", err)
+			}
+			data, err := txscript.PushedData(script)
+			if err != nil {
+				// XXX bad error, fix
+				return fmt.Errorf("   *** ERROR invalid "+
+					"script: %v", err)
+			}
+			if !bytes.Equal(data[0], flushRecord.Root[:]) {
+				continue
+			}
+
+			// Everything is cool so mark it and break out
+			done = true
+			break
+		}
+		if !done {
+			return fmt.Errorf("   *** ERROR merkle root not "+
+				"found: tx %v merkle %x", flushRecord.Tx,
+				flushRecord.Root)
+		}
+
+		// We are done
+		return nil
+	}
+
+	// 3. Make sure timestamps do not exist in the global database
+	for k := range digests {
+		key, err := hex.DecodeString(k)
+		if err != nil {
+			return fmt.Errorf("   *** ERROR internal error on "+
+				"key: %v", k)
+		}
+
+		if options.PrintHashes {
+			fmt.Printf("     Unflushed    : %v\n", k)
+		}
+
+		gdbts, err := fs.db.Get(key, nil)
+		if err != nil {
+			if err == leveldb.ErrNotFound {
+				continue
+			}
+			return fmt.Errorf("   *** ERROR found in db: %v %v",
+				k, err)
+		}
+		dbts := int64(binary.LittleEndian.Uint64(gdbts))
+		if dbts != ts {
```

Comment:
What about the scenario where `dbts == ts`?  

`dbts == ts` would mean that the global db is pointing to the correct timestamp for that digest, but the timestamp is missing a flush records so it shouldn't be in the global db in the first place.  The global db record should be deleted.

Reply:
I am not following this. The subtlety here is that way early in the function we exit if there is no flush record, therefore, we can't get into the condition you describe. Am I missing something?

Reply:
Where is the exit for when `len(digests) > 0 && flushRecord == nil`?  I'm not seeing it.

Reply:
That is in order to deal with a timestamp that was completely emptied after fixes around line 575. So a timestamp had let's say one entry and it was deleted as a timestamp mismatch then the map entry is deleted too. If we don't then we'd have to run fsck twice to clean up after that. It is a bit of hack to prevent that. Should I comment on it?